compiler/compile/OMRCompilation.cpp

```diff
@@ -2732,3 +2732,25 @@ OMR::Compilation::getSnippetsToBePatchedOnClassRedefinition()
    {
    return self()->cg()->getSnippetsToBePatchedOnClassRedefinition();
    }
+
+TR::Block *
+OMR::Compilation::insertNewFirstBlock()
+   {
+   TR::Node *oldBBStart = getStartTree()->getNode();
+   TR::Block *oldFirstBlock = getStartTree()->getNode()->getBlock();
+   TR::Node *glRegDeps=NULL;
+   if (oldBBStart->getNumChildren() == 1)
+      glRegDeps = oldBBStart->getChild(0);
+
+   TR::CFG *cfg = getFlowGraph();
+   TR::Block *newFirstBlock = TR::Block::createEmptyBlock(oldBBStart, self(), oldFirstBlock->getFrequency());
+
+   newFirstBlock->takeGlRegDeps(self(), glRegDeps);
```

Comment:
I think it actually works correctly. Here is an example of such transformation(we inserted empty block 3 in front of block 2, block 2 has not changed):
```
n102n     BBStart <block_11> (freq 3)                                                         [0x7f4fe3065fa0] bci=[-1,0,104] rc=0 vc=301 vn=- li=- udi=- nc=1
n104n       GlRegDeps ()                                                                      [0x7f4fe3066040] bci=[-1,0,104] rc=1 vc=301 vn=- li=2 udi=- nc=3 flg=0x20
n105n         aRegLoad edx   other<parm 2 Lsun/nio/fs/UnixPath;>[#424  Parm] [flags 0x40000107 0x0 ]  [0x7f4fe3066090] bci=[-1,0,104] rc=2 vc=301 vn=- li=2 udi=- nc=0
n106n         aRegLoad esi   file<parm 1 Lsun/nio/fs/UnixPath;>[#423  Parm] [flags 0x40000107 0x0 ] (SeenRealReference )  [0x7f4fe30660e0] bci=[-1,0,104] rc=2 vc=301 vn=- li=2 udi=- nc=0 flg=0x8000
n107n         aRegLoad eax   this<'this' parm Lsun/nio/fs/UnixException;>[#422  Parm] [flags 0x40000107 0x0 ]  [0x7f4fe3066130] bci=[-1,0,104] rc=2 vc=301 vn=- li=2 udi=- nc=0
n109n     goto --> block_2 BBStart at n9n                                                     [0x7f4fe30661d0] bci=[-1,0,104] rc=0 vc=301 vn=- li=- udi=- nc=1
n108n       GlRegDeps ()                                                                      [0x7f4fe3066180] bci=[-1,0,104] rc=1 vc=301 vn=- li=2 udi=- nc=3 flg=0x20
n105n         ==>aRegLoad
n106n         ==>aRegLoad
n107n         ==>aRegLoad
n103n     BBEnd </block_11> =====                                                             [0x7f4fe3065ff0] bci=[-1,0,104] rc=0 vc=301 vn=- li=- udi=- nc=0

n9n       BBStart <block_2> (freq 3) (cold)                                                   [0x7f4fe2f99a00] bci=[-1,0,104] rc=0 vc=301 vn=- li=2 udi=- nc=1
n58n        GlRegDeps ()                                                                      [0x7f4fe30651e0] bci=[-1,0,104] rc=1 vc=301 vn=- li=2 udi=- nc=3 flg=0x20
n59n          aRegLoad edx   other<parm 2 Lsun/nio/fs/UnixPath;>[#424  Parm] [flags 0x40000107 0x0 ]  [0x7f4fe3065230] bci=[-1,0,104] rc=3 vc=301 vn=- li=2 udi=- nc=0
n60n          aRegLoad esi   file<parm 1 Lsun/nio/fs/UnixPath;>[#423  Parm] [flags 0x40000107 0x0 ] (SeenRealReference )  [0x7f4fe3065280] bci=[-1,0,104] rc=3 vc=301 vn=- li=2 udi=- nc=0 flg=0x8000
n61n          aRegLoad eax   this<'this' parm Lsun/nio/fs/UnixException;>[#422  Parm] [flags 0x40000107 0x0 ]  [0x7f4fe30652d0] bci=[-1,0,104] rc=3 vc=301 vn=- li=2 udi=- nc=0
n15n      ifacmpne --> block_4 BBStart at n1n ()                                              [0x7f4fe2f99be0] bci=[-1,1,104] rc=0 vc=301 vn=- li=2 udi=- nc=3 flg=0x20
n60n        ==>aRegLoad
n12n        aconst NULL (X==0 )                                                               [0x7f4fe2f99af0] bci=[-1,1,104] rc=1 vc=301 vn=- li=2 udi=- nc=0 flg=0x2
n62n        GlRegDeps ()                                                                      [0x7f4fe3065320] bci=[-1,1,104] rc=1 vc=301 vn=- li=2 udi=- nc=3 flg=0x20
n59n          ==>aRegLoad
n60n          ==>aRegLoad
n61n          ==>aRegLoad
n10n      BBEnd </block_2> (cold)                                                             [0x7f4fe2f99a50] bci=[-1,1,104] rc=0 vc=301 vn=- li=2 udi=- nc=0

```
@vijaysun-omr could you please take  a look and let us know what you think? It's a rare case, but it's better to get it right. We need to insert an empty block at the beginning of the method since all of its blocks are cold.

Reply:
I am actually starting to have some doubts. For the purpose of this PR, we need to insert a new EBB at the start of the CFG. On the other hand, we need to convey resister/parameter association for the original start EBB (block_2). If the block remains empty, most likely things will work. But if the same insertion method is used somewhere else and then some code is added into block_3 there might be issues...

Reply:
So, I am assuming the `goto` is present because we need to separate that first block from the rest of the method that is all marked cold (in warm vs cold code cache).

Assuming this is right and the `goto` is needed, then the `GlRegDeps` state shown seems correct to me. i.e. we don't need any more or any fewer `GlRegDeps` and the commoning among the children also seems right to me.